cekit-helpers/cekit-cache-server/src/main/java/org/kie/cekit/cacher/utils/CacherUtils.java

```diff
@@ -70,6 +74,49 @@ public void cleanTmpFiles() {
         }
     }
 
+    /**
+     * Clean RHPAM and RHDM nightly artifacts, keep only latest 3 nightly builds
+     * run once a day.
+     */
+    @Scheduled(every = "24h", delay = 12, delayUnit = TimeUnit.HOURS)
+    public void cleanOldProductNightlyArtifacts() throws IOException, InterruptedException {
+        log.fine("Trying to identify the latest nightly build based on filesystem artifacts to delete, this could take a while...");
+        try {
+            log.fine("Walking through data dir searching for old nightly product builds...");
+
+            Map<File, LocalDate> nightlyBuildArtifacts = new HashMap<>();
+            List<LocalDate> foundDates = new ArrayList<>();
+            Files.walk(Paths.get(cacherProperties.getCacherArtifactsDir()))
+                    .map(Path::toFile)
+                    .filter(file -> cacherProperties.buildDatePattern.matcher(file.getName()).find())
+                    .filter(file -> file.getName().endsWith(".zip"))
+                    .forEach(file -> {
+                        Matcher test = cacherProperties.buildDatePattern.matcher(file.getName());
+                        if (test.find()) {
+                            LocalDate d = LocalDate.parse(test.group(0), cacherProperties.formatter);
+                            // collect all files that matches the build date pattern
+                            nightlyBuildArtifacts.put(file, d);
+                            // collect all build dates to get the newst one to be the base date.
+                            if (!foundDates.contains(d)) {
```

Comment:
I thing we do not need to collect all dates here. We can get all dates from nightlyBuildArtifacts map using `nightlyBuildArtifacts.values()` and to avoid duplication we can store the values as a Set e.g. `new HashSet<>(nightlyBuildArtifacts.values());`

Reply:
good point :)
update locally, testing.